ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/sink/io/Gcs.java

```diff
@@ -43,58 +48,60 @@ public Ndjson(Storage storage, long maxBytes, int maxMessages, Duration maxDelay
     }
 
     private final Storage storage;
+    private final Function<BlobInfo, CompletableFuture<Void>> batchCloseHook;
 
     private Write(Storage storage, long maxBytes, int maxMessages, Duration maxDelay,
-        String batchKeyTemplate) {
+        String batchKeyTemplate, Function<BlobInfo, CompletableFuture<Void>> batchCloseHook) {
       super(maxBytes, maxMessages, maxDelay, batchKeyTemplate);
       this.storage = storage;
+      this.batchCloseHook = batchCloseHook;
     }
 
     @Override
     protected String getBatchKey(PubsubMessage input) {
       return batchKeyTemplate.apply(input);
     }
 
+    private static final String BUCKET = "bucket";
+    private static final String NAME = "name";
+    private static final Pattern blobIdPattern = Pattern
+        .compile("(gs://)?(?<" + BUCKET + ">[^/]+)(/(?<" + NAME + ">.*))?");
+
     @Override
     protected Batch getBatch(String gcsPrefix) {
-      return new Batch(storage, gcsPrefix);
+      final Matcher gcsPrefixMatcher = blobIdPattern.matcher(gcsPrefix);
+      if (!gcsPrefixMatcher.matches()) {
+        throw new IllegalArgumentException("Gcs prefix must match \"" + blobIdPattern.pattern()
+            + "\" but got \"" + gcsPrefix + "\" from: " + batchKeyTemplate.template);
```

Comment:
This might be easier to read with `String.format`

```java
         throw new IllegalArgumentException(String.format("Gcs prefix must match \"%s\" but got \"%s\" from: %s",
           blobIdPattern.pattern(), gcsPrefix, batchKeyTemplate.template));
```